HW-4 Data Binding(+Recent Search)

juan-rybczinski opened the pull request:
- 검색이 시작되고 끝날 때 ProgressBar visibility를 Data Binding으로 제어해 보려고 했는데 잘 안되네요.(Binding Adapter를 써야 할까요?)
- ChipGroup과 Chip을 활용해 봤는데 Chip 구성할 때 Data Binding을 어떻게 활용해야할 지 잘 몰라서 그냥 View에서 로직으로 만들어 버렸습니다 ;
- Coroutine을 쓰고는 있지만 잘 몰라서 ; Coroutine 안에서 반환 받은 값을 Binding Adapter에서 쓰려고 했더니 시점이 어긋나 버리네요.
- 공통되는 것 같은 로직을 Base로 하나하나 빼다보니 Base로 거의 다 옮겨 버렸네요 ;

juan-rybczinski committed:
HW-4 Fix minor bugs
- Set layout_height of item_movie root view to wrap_content
- Pass movie list to the constructor of MovieViewHolder

juan-rybczinski committed:
HW-4 Add some features before assignment
- Format title regarding HTML
- Add ProgressBar
- Add RatingBar

juan-rybczinski committed:
HW-4 Replace ViewBinding with DataBinding

juan-rybczinski committed:
HW-4 Apply basic DataBinding on fragment_movie and item_movie

juan-rybczinski committed:
HW-4 Apply BindingAdapter
- Format title with HTML
- Loading image with placeholder and error
- Try to use DataBinding on click listener of ViewHolder

juan-rybczinski committed:
HW-4 Add RecentFragment
- Add RecentContract, RecentPresenter, RecentFragment
- Add Recent button in fragment_movie and navigate to fragment_recent when clicked
- Move movieRepository declaration to BaseFragment

juan-rybczinski committed:
HW-4 Save and retrieve RecentSearch
- Save the search keyword whenever the search is done
- Retrieve 5 recent keywords when recent button is clicked

juan-rybczinski committed:
HW-4 Implement to search by recent keywords
- Visualize recent keywords by ChipGroup and Chip

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/BaseFragment.kt:
```diff
@@ -11,20 +11,41 @@ import android.view.inputmethod.InputMethodManager
+
     protected abstract val presenter: P
 
+    protected val movieRepository: MovieRepository by lazy {
```
repository 관리는 Fragment 가아니라 Prensenter에서 하지않나용 ????

juan-rybczinski replied:
Repository에 접근하는 로직은 Presenter에 다 있으니 Repository는 Presenter로 가는게 맞겠네요. Presenter에 Repository 대신 Context를 전달하게 바꾸었습니다. MVP 역할 분리가 완전하지 않았는데 좋은 지적 감사합니다!

ab19c9ef5

hunihun replied:
:+1:

hunihun commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/BaseFragment.kt:
```diff
@@ -11,20 +11,41 @@ import android.view.inputmethod.InputMethodManager
         requireActivity().getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
     }
 
+    protected val movieAdapter = MovieAdapter {
```
BaseFragment 에 Adapter까지 관리할 필요가있나용 ???
뭔가 Base의 성격에 안맞는 듯한 느낌..?? (ProgressBar or Toast 등등이 맞는것같습니다!)

juan-rybczinski replied:
확실히 Base에 movieAdapter가 있는 건 이상하네요 😂
하지만 Movie와 Recent의 공통 로직은 공유하고 싶어 MovieList라는 새로운 추상화 계층을 또 만들었습니다. 이렇게 해도 되겠죠? ㅋㅋㅋ

1b3f84878

hunihun replied:
MVP 패턴에 대한 단점중 view와 present 가 1:1 구조이기 때문에 중복되는 로직에 대한 문제를 본것같아요!
그걸 해결하기 위한게 MVVM이라고 알고있는데 MVP를 사용할때 중복로직에 대한 베스트가 뭔지 잘 모르겠네요..ㅎㅎ

hunihun reviewed (commented):
늦어서 죄송합니다 ㅠㅠ (아직 제것도 못했어용...) 
프로젝트가 점점 커져서 어렵네요 ...ㅎㅎ, Room쪽은 제가 아예 몰라서 넘겼습니다 ㅠㅠ

juan-rybczinski committed:
HW-4 Apply executePendingBindings on RecyclerView

juan-rybczinski committed:
HW-4 Add MovieList abstract layer to share common functions between Movie and Recent
- Move movieAdapter from Base to MovieList(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/915#discussion_r530954343)

juan-rybczinski committed:
HW-4 Move repository from View to Presenter(https://github.com/StudyFork/GoogryAndroidArchitectureStudy/pull/915#discussion_r530951438)

hunihun commented:
고생하셨습니다!!

juan-rybczinski committed:
HW-4 Change Presenter to use context inside View passed in

sjjeong commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/network/MovieHeaderInterceptor.kt:
```diff
@@ -11,7 +11,10 @@ class MovieHeaderInterceptor : Interceptor {
-                .addHeader("X-Naver-Client-Secret", App.instance.getString(R.string.naver_client_secret))
+                .addHeader(
+                    "X-Naver-Client-Secret",
+                    App.instance.getString(R.string.naver_client_secret)
```
key 값을 string res보다 kotlin에서 상수로 관리하게 해주세요

juan-rybczinski replied:
BuildConfig 활용하게 수정했습니다.
gitignore에 등록한 secrets.xml로 관리하는 방법은 추천하지 않으시는 건가요?

 e13ef6af9

sjjeong replied:
resource에 접근하려면 context가 필요해서 저는 잘 안하는 편입니다.
라이브러리 내부에서 resource key로 접근하면 어쩔수 없이 그렇게 하지만 제가 관리하는 거라면 kotlin에서 관리해요

sjjeong commented on class14/juan-rybczinski/GoogryAndroidArchitectureStudy/app/src/main/java/com/example/googryandroidarchitecturestudy/ui/fragment/BaseFragment.kt:
```diff
@@ -13,13 +13,14 @@ import androidx.viewbinding.ViewBinding
     protected val binding get() = _binding!!
 
-    protected abstract val presenter: P
+    var onProgress = View.INVISIBLE
```
이거 프로그래스가 실제로는 동작을 안할 것 같은데 mvvm하면서 개선 될 거니까 넘어갈게요

juan-rybczinski committed:
HW-4 Change to use BuildConfig to hold secret information

sjjeong reviewed (approved):
LGTM ✊ ✌️ 👋
